File: jaxlib/mosaic/python/infer_memref_layout.py
# Copyright 2023 The JAX Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Inference for memref layout and memory space."""

# mypy: ignore-errors
from jaxlib.mlir import ir
from jaxlib.mlir.dialects import func
import numpy as np

from . import tpu


def _round_up(x: int, to: int):
  return ((x + to - 1) // to) * to


def _tiling_factor(
    num_128s: int, hardware_generation: int, bitwidth: int
) -> int:
  """Returns the number of 128-element groups in a tile.

  Arguments:
    num_128s: A number of 128-element groups in the full operand.
    hardware_generation: An integer indicating the target TPU generation.
    bitwidth: The bitwidth of the element type of the operand.
  """
  if num_128s == 1 and hardware_generation >= 4 and bitwidth == 32:
    return 1
  elif num_128s <= 2 and (
      bitwidth == 32 or (hardware_generation >= 4 and bitwidth == 16)
  ):
    return 2
  elif num_128s <= 4:
    return 4
  else:
    return 8


def infer_memref(
    memref: ir.MemRefType, hardware_generation: int
) -> ir.MemRefType:
  """Infers the layout and memory space attributes of the given memref.

  Arguments:
    memref: The memref type with potentially missing layout and memory space.
    hardware_generation: The TPU hardware generation to target.

  Returns:
    A memref type with layout and memory space filled in.
  """
  vmem = ir.Attribute.parse("#tpu.memory_space<vmem>")
  if tpu.private_has_no_memory_space(memref):
    memory_space = vmem
  else:
    memory_space = memref.memory_space
  if ir.AffineMapAttr.isinstance(memref.layout):
    if not tpu.private_is_identity(memref.layout):
      raise NotImplementedError("Non-identity affine layout")
    bitwidth: int
    if ir.BF16Type.isinstance(memref.element_type):
      bitwidth = 16
    elif ir.F32Type.isinstance(memref.element_type):
      bitwidth = 32
    elif ir.IntegerType.isinstance(memref.element_type):
      bitwidth = ir.IntegerType(memref.element_type).width
    else:
      raise NotImplementedError(
          f"Unrecognized element type: {memref.element_type}"
      )
    # Infer the layout.
    if memref.rank == 1:
      tile = _tiling_factor(
          ((memref.shape[-1] + 127) // 128), hardware_generation, bitwidth
      ) * 128
      if bitwidth == 32:
        trailing_tiles = ""
      else:
        if bitwidth.bit_count() != 1 or bitwidth > 32:
          raise NotImplementedError(f"Unsupported bitwidth: {bitwidth}")
        trailing_tiles = f"(128)({32 // bitwidth},1)"
      layout = ir.Attribute.parse(
          f"#tpu.tiled<{memref.rank},({tile}){trailing_tiles}>"
      )
    else:
      leading_tile = _tiling_factor(
          memref.shape[-2], hardware_generation, bitwidth
      )
      if bitwidth == 32:
        trailing_tiles = ""
      else:
        if bitwidth.bit_count() != 1 or bitwidth > 32:
          raise NotImplementedError(f"Unsupported bitwidth: {bitwidth}")
        trailing_tiles = f"({32 // bitwidth},1)"
      layout = ir.Attribute.parse(
          f"#tpu.tiled<{memref.rank},({leading_tile},128){trailing_tiles}>"
      )
  elif tpu.private_is_tiled_layout(memref.layout):
    layout = memref.layout
  else:
    raise NotImplementedError("Unrecognized layout annotation")
  new_shape = list(memref.shape)
  # Make sure only the first tile might introduce padding.
  first_tile, *_ = tiles = tpu.private_get_tiles(layout)
  tiled_dims = np.asarray(first_tile, dtype=np.int32)
  for t in tiles:
    t = np.asarray(t, dtype=np.int32)
    if len(t) > len(tiled_dims):
      raise NotImplementedError("Layout too complicated")
    untiled_prefix, tiled_suffix = tiled_dims[:-len(t)], tiled_dims[-len(t):]
    if np.any(tiled_suffix % t != 0):
      raise NotImplementedError("Layout too complicated")
    tiled_dims = np.concatenate([untiled_prefix, tiled_suffix // t, t])
  untiled_dims = len(new_shape) - len(first_tile)
  if untiled_dims < 0:
    raise ValueError("Invalid tiling")
  for i, size in enumerate(first_tile):
    new_shape[untiled_dims + i] = _round_up(
        new_shape[untiled_dims + i], to=size
    )
  return ir.MemRefType.get(
      new_shape, memref.element_type, layout, memory_space
  )


def infer_module(module: ir.Module, hardware_generation: int):
  """Infers the layout and memory space attributes of function memref arguments.

  In the future we should require those annotations from Mosaic users, but it's
  best to keep them internal for as long as they are under development.

  Arguments:
    module: The MLIR module on which to perform the inference.
    hardware_generation: The TPU hardware generation to target.
  """
  # TODO(apaszke): Do layout assignment for scoped allocations too.
  for f in module.body:
    assert isinstance(f, func.FuncOp)
    if len(f.body.blocks) != 1:
      raise ValueError("Functions should only have a single block")
    (entry,) = f.body.blocks
    new_arg_types = []
    with ir.InsertionPoint.at_block_begin(entry):
      for arg in entry.arguments:
        try:
          memref = ir.MemRefType(arg.type)
        except ValueError:
          new_arg_types.append(arg.type)
          continue
        new_memref = infer_memref(memref, hardware_generation)
        arg.set_type(new_memref)
        new_arg_types.append(new_memref)
        if memref != new_memref:
          # Some standard MLIR ops have static checks that seems unreasonable,
          # and we know they hold in the way they are used in Mosaic. Still,
          # verification with layouts likes to fail, because it can't statically
          # prove the properties.
          erase_op = tpu.EraseLayoutOp(
              ir.MemRefType.get(
                  new_memref.shape, memref.element_type,
                  None, new_memref.memory_space
              ),
              arg,
          )
          tpu.private_replace_all_uses_except(arg, erase_op.result, erase_op)
      f.attributes["function_type"] = ir.TypeAttr.get(
          ir.FunctionType.get(new_arg_types, f.type.results)
      )

File: jaxlib/mosaic/dialect/tpu/transforms/infer_memref_layout.cc
#include "jaxlib/mosaic/dialect/tpu/transforms/infer_memref_layout.h"

#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <memory>

#include "llvm/ADT/bit.h"
#include "llvm/Support/MathExtras.h"
#include "mlir/Dialect/Func/IR/FuncOps.h"
#include "mlir/Dialect/MemRef/IR/MemRef.h"
#include "mlir/IR/Attributes.h"
#include "mlir/IR/Block.h"
#include "mlir/IR/Builders.h"
#include "mlir/IR/BuiltinAttributes.h"
#include "mlir/IR/BuiltinOps.h"
#include "mlir/IR/BuiltinTypes.h"
#include "mlir/IR/Diagnostics.h"
#include "mlir/IR/Location.h"
#include "mlir/IR/Region.h"
#include "mlir/IR/Value.h"
#include "mlir/Pass/Pass.h"
#include "mlir/Support/LLVM.h"
#include "mlir/Support/LogicalResult.h"
#include "absl/log/check.h"
#include "jaxlib/mosaic/dialect/tpu/tpu_dialect.h"
#include "jaxlib/mosaic/dialect/tpu/util.h"
#include "xla/layout.h"

namespace mlir::tpu {

#define GEN_PASS_DECL_INFERMEMREFLAYOUTPASS
#define GEN_PASS_DEF_INFERMEMREFLAYOUTPASS
#include "jaxlib/mosaic/dialect/tpu/tpu_passes.h.inc"


// Returns the number of 128-element groups in a tile.
//
// Arguments:
//   num_128s: A number of 128-element groups in the full operand.
//   hardware_generation: An integer indicating the target TPU generation.
//   tpu_tiling_flags: A struct of flags indicating which large tiling modes are
//     enabled by XLA for memrefs.
//   bitwidth: The bitwidth of the element type of the operand.
int getTilingFactor(const int num_128s, const int hardware_generation,
                    const TpuTilingFlags &tpu_tiling_flags,
                    const int8_t bitwidth) {
  CHECK(llvm::isPowerOf2_32(bitwidth));
  CHECK_LE(4, bitwidth);
  CHECK_LE(bitwidth, 32);
  const int packing = 32 / bitwidth;
  const int min_tiling = (1 + (hardware_generation < 4)) * packing;
  const int max_normal_tiling = 8;

  const int large_tiling = [&] {
    if (bitwidth == 4 && tpu_tiling_flags.use_x4_large_second_minor) {
      return 64;
    }
    if (bitwidth == 8 && tpu_tiling_flags.use_x8_large_second_minor) {
      return 32;
    }
    if (bitwidth == 16 && tpu_tiling_flags.use_x16_large_second_minor) {
      return 16;
    }
    return 8;
  }();

  // Use large tiling if our operand is tall enough to fit at least one full
  // tile.
  if (large_tiling <= num_128s) {
    return large_tiling;
  }

  int tiling = min_tiling;
  while (tiling < std::min(num_128s, max_normal_tiling)) {
    tiling *= 2;
  }
  return tiling;
}

FailureOr<TiledLayoutAttr> inferLayout(MemRefType memref_ty,
                                       const int hardware_generation,
                                       const TpuTilingFlags &tpu_tiling_flags,
                                       int64_t leading_tile_rows = 0) {
  if (auto tiled_layout_attr =
          dyn_cast<TiledLayoutAttr>(memref_ty.getLayout())) {
    return tiled_layout_attr;
  }
  if (auto affine_map_attr = dyn_cast<AffineMapAttr>(memref_ty.getLayout())) {
    if (memref_ty.getRank() == 0) {
      return emitError(UnknownLoc::get(memref_ty.getContext()),
                       "0-rank memref not supported");
    }
    if (!affine_map_attr.isIdentity()) {
      return emitError(UnknownLoc::get(memref_ty.getContext()),
                       "Non-identity affine layout");
    }
    if (!memref_ty.getElementType().isIntOrFloat()) {
      return emitError(UnknownLoc::get(memref_ty.getContext()),
                       "Invalid element type for memref");
    }
    const int8_t bitwidth = memref_ty.getElementTypeBitWidth();
    // Infer the layout
    if (memref_ty.getRank() == 1) {
      const int64_t leading_tile =
          getTilingFactor(llvm::divideCeil(memref_ty.getShape().back(), 128),
                          hardware_generation, tpu_tiling_flags, bitwidth) *
          128;
      SmallVector<xla::Tile> tiles{xla::Tile({leading_tile})};
      if (bitwidth != 32) {
        if (!llvm::has_single_bit<unsigned>(bitwidth) || bitwidth > 32) {
          return emitError(UnknownLoc::get(memref_ty.getContext()),
                           "Unsupported bitwidth: ")
                 << bitwidth;
        }
        tiles.append({xla::Tile({128}), xla::Tile({32 / bitwidth, 1})});
      }
      return TiledLayoutAttr::get(memref_ty.getContext(), tiles, {1});
    }

    // memref.getRank() > 1
    const ArrayRef<int64_t> shape = memref_ty.getShape();
    const int64_t second_minor = shape[shape.size() - 2];
    if (leading_tile_rows == 0) {
      leading_tile_rows = getTilingFactor(second_minor, hardware_generation,
                                          tpu_tiling_flags, bitwidth);
    }
    SmallVector<xla::Tile> tiles{xla::Tile({leading_tile_rows, 128})};
    if (bitwidth != 32) {
      if (!llvm::has_single_bit<unsigned>(bitwidth) || bitwidth > 32) {
        return emitError(UnknownLoc::get(memref_ty.getContext()),
                         "Unsupported bitwidth: ")
               << bitwidth;
      }
      tiles.push_back(xla::Tile({32 / bitwidth, 1}));
    }
    auto tile_strides = ComputeTileStrides(memref_ty, {leading_tile_rows, 128});
    return TiledLayoutAttr::get(memref_ty.getContext(), tiles, tile_strides);
  }
  return emitError(UnknownLoc::get(memref_ty.getContext()),
                   "Unrecognized layout annotation");
}

// Make sure only the first tile might introduce padding.
LogicalResult checkTiles(MLIRContext *mlir_ctx,
                         const ArrayRef<xla::Tile> &tiles) {
  SmallVector<int64_t> tiled_dims(tiles.front().dimensions().begin(),
                                  tiles.front().dimensions().end());
  for (const xla::Tile &t : tiles.drop_front()) {
    const int64_t offset = tiled_dims.size() - t.dimensions().size();
    if (offset < 0) {
      return emitError(UnknownLoc::get(mlir_ctx),
                       "Not implemented: layout too complicated");
    }
    for (int i = 0; i < t.dimensions().size(); ++i) {
      auto [d, m] = std::div(tiled_dims[offset + i], t.dimension(i));
      if (m != 0) {
        return emitError(UnknownLoc::get(mlir_ctx),
                         "Not implemented: layout too complicated");
      }
      tiled_dims[offset + i] = d;
    }
    tiled_dims.append(t.dimensions().begin(), t.dimensions().end());
  }
  return success();
}

FailureOr<MemRefType> inferMemref(MemRefType memref,
                                  const int hardware_generation,
                                  const TpuTilingFlags &tpu_tiling_flags,
                                  int64_t leading_tile_rows) {
  if (isa<SemaphoreType, DMASemaphoreType>(memref.getElementType())) {
    const Attribute semaphore_mem = tpu::MemorySpaceAttr::get(
        memref.getContext(), MemorySpace::kSemaphoreMem);
    SmallVector<int64_t> tile_strides;
    tile_strides.reserve(memref.getRank());
    int64_t stride = 1;
    for (int i = memref.getRank() - 1; i >= 0; --i) {
      tile_strides.push_back(stride);
      stride *= memref.getDimSize(i);
    }
    std::reverse(tile_strides.begin(), tile_strides.end());
    auto layout = TiledLayoutAttr::get(memref.getContext(), {}, tile_strides);
    return MemRefType::get(memref.getShape(), memref.getElementType(), layout,
                           semaphore_mem);
  }
  const Attribute vmem =
      tpu::MemorySpaceAttr::get(memref.getContext(), MemorySpace::vmem);
  const Attribute memory_space =
      memref.getMemorySpace() == nullptr ? vmem : memref.getMemorySpace();
  FAILUREOR_ASSIGN_OR_RETURN(const TiledLayoutAttr layout,
                             inferLayout(memref, hardware_generation,
                                         tpu_tiling_flags, leading_tile_rows));

  const ArrayRef<xla::Tile> tiles = layout.getTiles();
  if (failed(checkTiles(memref.getContext(), tiles))) {
    return failure();
  }
  const xla::Tile &first_tile = tiles.front();
  const int64_t untiled_dims =
      memref.getShape().size() - first_tile.dimensions().size();
  if (untiled_dims < 0) {
    return emitError(UnknownLoc::get(memref.getContext()), "Invalid tiling");
  }
  SmallVector<int64_t> new_shape(memref.getShape());
  for (int i = 0; i < first_tile.dimensions().size(); ++i) {
    new_shape[untiled_dims + i] =
        llvm::alignTo(new_shape[untiled_dims + i], first_tile.dimension(i));
  }
  return MemRefType::get(new_shape, memref.getElementType(), layout,
                         memory_space);
}

LogicalResult inferOp(Operation &op, const int hardware_generation,
                      const TpuTilingFlags &tpu_tiling_flags) {
  if (auto alloca_op = dyn_cast<memref::AllocaOp>(op)) {
    TypedValue<MemRefType> arg = alloca_op.getResult();
    const MemRefType memref_ty = alloca_op.getResult().getType();
    FAILUREOR_ASSIGN_OR_RETURN(
        const MemRefType new_memref_ty,
        inferMemref(memref_ty, hardware_generation, tpu_tiling_flags));
    alloca_op.getResult().setType(new_memref_ty);
    if (memref_ty != new_memref_ty) {
      OpBuilder builder(alloca_op->getContext());
      builder.setInsertionPointAfter(alloca_op);
      auto erase_op = builder.create<tpu::EraseLayoutOp>(
          arg.getLoc(),
          MemRefType::get(new_memref_ty.getShape(), memref_ty.getElementType(),
                          /*layout=*/nullptr, new_memref_ty.getMemorySpace()),
          arg);
      arg.replaceAllUsesExcept(erase_op.getResult(), erase_op);
    }
  } else if (auto alloca_op = dyn_cast<tpu::AllocaSemaphoreOp>(op)) {
    TypedValue<MemRefType> arg = alloca_op.getResult();
    const MemRefType memref_ty = alloca_op.getResult().getType();
    FAILUREOR_ASSIGN_OR_RETURN(
        const MemRefType new_memref_ty,
        inferMemref(memref_ty, hardware_generation, tpu_tiling_flags));
    alloca_op.getResult().setType(new_memref_ty);
    if (memref_ty != new_memref_ty) {
      OpBuilder builder(alloca_op->getContext());
      builder.setInsertionPointAfter(alloca_op);
      auto erase_op = builder.create<tpu::EraseLayoutOp>(
          arg.getLoc(),
          MemRefType::get(new_memref_ty.getShape(), memref_ty.getElementType(),
                          /*layout=*/nullptr, new_memref_ty.getMemorySpace()),
          arg);
      arg.replaceAllUsesExcept(erase_op.getResult(), erase_op);
    }
  }
  for (Region &region : op.getRegions()) {
    for (Block &block : region) {
      for (Operation& op : block) {
        if (failed(inferOp(op, hardware_generation, tpu_tiling_flags))) {
          return failure();
        }
      }
    }
  }
  return success();
}

LogicalResult inferFunc(func::FuncOp f, const int hardware_generation,
                        const TpuTilingFlags &tpu_tiling_flags) {
  if (!f.getBody().hasOneBlock()) {
    return f.emitOpError("Functions should only have a single block");
  }
  Block &entry = f.getBody().front();
  SmallVector<Type> new_arg_types;
  auto builder = OpBuilder::atBlockBegin(&entry);
  for (int i = 0; i < entry.getNumArguments(); ++i) {
    BlockArgument arg = entry.getArgument(i);
    const auto memref_ty = dyn_cast<MemRefType>(arg.getType());
    if (memref_ty == nullptr) {
      new_arg_types.push_back(arg.getType());
      continue;
    }
    int64_t leading_tile_rows = 0;
    auto leading_tile_rows_attr =
        f.getArgAttrOfType<mlir::IntegerAttr>(i, kLeadingTileRows);
    if (leading_tile_rows_attr != nullptr) {
      leading_tile_rows = leading_tile_rows_attr.getInt();
      f.removeArgAttr(i, kLeadingTileRows);
    }

    FAILUREOR_ASSIGN_OR_RETURN(
        const MemRefType new_memref_ty,
        inferMemref(memref_ty, hardware_generation, tpu_tiling_flags,
                    leading_tile_rows));
    arg.setType(new_memref_ty);
    new_arg_types.push_back(arg.getType());
    if (memref_ty != new_memref_ty) {
      // Some standard MLIR ops have static checks that seems unreasonable,
      // and we know they hold in the way they are used in Mosaic. Still,
      // verification with layouts likes to fail, because it can't statically
      // prove the properties.
      auto erase_op = builder.create<tpu::EraseLayoutOp>(
          arg.getLoc(),
          MemRefType::get(new_memref_ty.getShape(), memref_ty.getElementType(),
                          /*layout=*/nullptr, new_memref_ty.getMemorySpace()),
          arg);
      arg.replaceAllUsesExcept(erase_op.getResult(), erase_op);
    }
  }
  f.setFunctionType(
      builder.getAttr<FunctionType>(new_arg_types, f.getResultTypes()));
  for (Operation &op : entry.getOperations()) {
    if (failed(inferOp(op, hardware_generation, tpu_tiling_flags))) {
      return failure();
    }
  }
  return success();
}

// Infers the layout and memory space attributes of function memref arguments.
//
// In the future we should require those annotations from Mosaic users, but it's
// best to keep them internal for as long as they are under development.
//
// Arguments:
//   module: The MLIR module on which to perform the inference.
//   hardware_generation: The TPU hardware generation to target.
LogicalResult inferModule(ModuleOp module, const int hardware_generation,
                          const TpuTilingFlags &tpu_tiling_flags) {
  // TODO(apaszke): Do layout assignment for scoped allocations too.
  for (Operation &op : *module.getBody()) {
    auto f = dyn_cast<func::FuncOp>(op);
    if (f == nullptr) {
      return module.emitOpError("Expected only FuncOps but found ") << op;
    }
    if (failed(inferFunc(f, hardware_generation, tpu_tiling_flags))) {
      return failure();
    }
  }
  return success();
}

struct InferMemRefLayoutPass
    : public impl::InferMemRefLayoutPassBase<InferMemRefLayoutPass> {
  InferMemRefLayoutPass(int hardware_generation_,
                        const TpuTilingFlags &tpu_tiling_flags_) {
    hardware_generation = hardware_generation_;
    tpu_tiling_flags = tpu_tiling_flags_;
  }
  void runOnOperation() override {
    // Fail if hardware_generation has not been set from the default value.
    if (hardware_generation < 0) {
      signalPassFailure();
      return;
    }
    func::FuncOp func = getOperation();
    if (failed(inferFunc(func, hardware_generation, tpu_tiling_flags))) {
      signalPassFailure();
      return;
    }
  }
};

std::unique_ptr<OperationPass<func::FuncOp>> createInferMemRefLayoutPass(
    int hardware_generation, const TpuTilingFlags &tpu_tiling_flags_) {
  return std::make_unique<InferMemRefLayoutPass>(hardware_generation,
                                                 tpu_tiling_flags_);
}

}  // namespace mlir::tpu
